File: Server/logger/config/resolve_logger_config.py
"""Resolve logger config with environment-aware defaults."""

from typing import Any

from .config_model import LoggerConfig
from .logger_contract_generated import LOGGER_BOOLEAN_KEYS, build_environment_defaults
from .logger_env import (
    is_logger_production,
    is_logger_verbose_dev,
    parse_dev_category_overrides,
)


def apply_dev_category_overrides(config: dict[str, Any], keys: list[str]) -> dict[str, Any]:
    if not keys:
        return config
    next_config = dict(config)
    for key in keys:
        if key == "api" or key in LOGGER_BOOLEAN_KEYS:
            next_config[key] = True
    return next_config


def apply_dev_verbose(config: dict[str, Any]) -> dict[str, Any]:
    next_config = dict(config)
    for key in LOGGER_BOOLEAN_KEYS:
        next_config[key] = True
    next_config["api"] = True
    next_config["logLevel"] = "DEBUG"
    next_config["errors"] = True
    next_config["security"] = True
    return next_config


def apply_production_guard(config: dict[str, Any]) -> dict[str, Any]:
    next_config = dict(config)
    for key in LOGGER_BOOLEAN_KEYS:
        next_config[key] = True
    next_config["api"] = True
    next_config["errors"] = True
    next_config["security"] = True
    return next_config


def resolve_logger_config(overrides: dict[str, Any] | None = None) -> LoggerConfig:
    is_prod = is_logger_production()
    config_dict = build_environment_defaults(is_prod)
    if overrides:
        config_dict.update(overrides)

    if not is_prod:
        if is_logger_verbose_dev():
            config_dict = apply_dev_verbose(config_dict)
        else:
            config_dict = apply_dev_category_overrides(config_dict, parse_dev_category_overrides())
    else:
        config_dict = apply_production_guard(config_dict)

    return LoggerConfig(config_dict)


def merge_logger_config_update(current: LoggerConfig, updates: dict[str, Any]) -> LoggerConfig:
    merged = current.to_dict()
    merged.update(updates)
    config = LoggerConfig(merged)
    if is_logger_production():
        guarded = apply_production_guard(config.to_dict())
        return LoggerConfig(guarded)
    return config

File: Client/packages/logger/config/resolveLoggerConfig.ts
import type {
  ApiSubcategoryConfig,
  LoggerConfig,
  LogLevel,
} from "packages/logger/core/loggerTypes";

import {
  buildEnvironmentDefaults,
  buildProductionApiConfig,
  DEFAULT_API_CONFIG_EXPORT,
  LOGGER_BOOLEAN_KEYS,
  type LoggerBooleanKey,
} from "./loggerContract.generated";
import { isLoggerProduction, isLoggerVerboseDev, parseDevCategoryOverrides } from "./loggerEnv";

export { LOGGER_BOOLEAN_KEYS, type LoggerBooleanKey };

const DEFAULT_API_CONFIG: ApiSubcategoryConfig = DEFAULT_API_CONFIG_EXPORT;

function mergeApiConfig(
  base: LoggerConfig["api"],
  override: LoggerConfig["api"] | undefined
): LoggerConfig["api"] {
  if (override === undefined) {
    return base;
  }
  if (typeof override === "boolean") {
    if (override) {
      return buildProductionApiConfig();
    }
    return { ...DEFAULT_API_CONFIG };
  }
  if (typeof base === "boolean") {
    return override;
  }
  return { ...base, ...override };
}

function applyDevCategoryOverrides(config: LoggerConfig, keys: string[]): LoggerConfig {
  if (keys.length === 0) {
    return config;
  }

  const next = { ...config };
  for (const key of keys) {
    if (key === "api") {
      next.api = buildProductionApiConfig();
      continue;
    }
    if (LOGGER_BOOLEAN_KEYS.includes(key as LoggerBooleanKey)) {
      (next as Record<string, unknown>)[key] = true;
    }
  }
  return next;
}

function applyDevVerbose(config: LoggerConfig): LoggerConfig {
  const next = { ...config };
  for (const key of LOGGER_BOOLEAN_KEYS) {
    if (key === "errors" || key === "security") {
      continue;
    }
    (next as Record<string, unknown>)[key] = true;
  }
  next.api = buildProductionApiConfig();
  next.logLevel = "DEBUG";
  return next;
}

export function applyProductionGuard(config: LoggerConfig): LoggerConfig {
  const next = { ...config };
  for (const key of LOGGER_BOOLEAN_KEYS) {
    (next as Record<string, unknown>)[key] = true;
  }
  next.errors = true;
  next.security = true;
  next.api = buildProductionApiConfig();
  return next;
}

export { buildEnvironmentDefaults };

export function resolveLoggerConfig(overrides?: Partial<LoggerConfig>): LoggerConfig {
  const isProd = isLoggerProduction();
  let config = buildEnvironmentDefaults(isProd);

  if (overrides) {
    config = {
      ...config,
      ...overrides,
      api: mergeApiConfig(config.api, overrides.api),
    };
  }

  if (!isProd) {
    if (isLoggerVerboseDev()) {
      config = applyDevVerbose(config);
    } else {
      config = applyDevCategoryOverrides(config, parseDevCategoryOverrides());
    }
  } else {
    config = applyProductionGuard(config);
  }

  return config;
}

export function mergeLoggerConfigUpdate(
  current: LoggerConfig,
  updates: Partial<LoggerConfig>
): LoggerConfig {
  const merged: LoggerConfig = {
    ...current,
    ...updates,
    api: mergeApiConfig(current.api, updates.api),
  };

  if (isLoggerProduction()) {
    return applyProductionGuard(merged);
  }

  return merged;
}

export function normalizeLogLevel(value: unknown, fallback: LogLevel): LogLevel {
  if (value === "DEBUG" || value === "INFO" || value === "WARN" || value === "ERROR") {
    return value;
  }
  return fallback;
}
